src/app/3d/qgs3dmeasuredialog.cpp

```diff
@@ -0,0 +1,385 @@
+  buttonBox->addButton( cb, QDialogButtonBox::ActionRole );
+  connect( cb, &QAbstractButton::clicked, this, &Qgs3DMeasureDialog::openConfigTab );
+
+  repopulateComboBoxUnits();
+
+  // Choose unit
+  if ( mUseMapUnits )
+    mUnitsCombo->setCurrentIndex( mUnitsCombo->findData( QgsUnitTypes::DistanceUnknownUnit ) );
+  else
+    mUnitsCombo->setCurrentIndex( mUnitsCombo->findData( QgsProject::instance()->distanceUnits() ) );
+
+  QgsMapCanvas *canvas2D = QgisApp::instance()->mapCanvas();
+
+  updateSettings();
+
+  qInfo() << "3D Measure Dialog created";
+  connect( buttonBox, &QDialogButtonBox::rejected, this, &Qgs3DMeasureDialog::reject );
+  connect( mUnitsCombo, static_cast<void ( QComboBox::* )( int )>( &QComboBox::currentIndexChanged ), this, &Qgs3DMeasureDialog::unitsChanged );
+  connect( buttonBox, &QDialogButtonBox::helpRequested, this, &Qgs3DMeasureDialog::showHelp );
+  connect( canvas2D, &QgsMapCanvas::destinationCrsChanged, this, &Qgs3DMeasureDialog::crsChanged );
+}
+
+void Qgs3DMeasureDialog::saveWindowLocation()
+{
+  QgsSettings settings;
+  settings.setValue( QStringLiteral( "Windows/3DMeasure/geometry" ), saveGeometry() );
+  const QString &key = "/Windows/3DMeasure/h";
+  settings.setValue( key, height() );
+}
+
+void Qgs3DMeasureDialog::restorePosition()
+{
+  QgsSettings settings;
+  restoreGeometry( settings.value( QStringLiteral( "Windows/3DMeasure/geometry" ) ).toByteArray() );
+  int wh = settings.value( QStringLiteral( "Windows/3DMeasure/h" ), 200 ).toInt();
+  resize( width(), wh );
+  updateUi();
+}
+
+void Qgs3DMeasureDialog::addPoint()
+{
+  int numPoints = mTool->points().size();
+  qInfo() << "Add point. Current num points: " << numPoints;
+  if ( numPoints > 1 )
+  {
+    if ( !mTool->done() )
+    {
+      QTreeWidgetItem *item = new QTreeWidgetItem( QStringList( QLocale().toString( 0.0, 'f', mDecimalPlaces ) ) );
+      item->setTextAlignment( 0, Qt::AlignRight );
+      mTable->addTopLevelItem( item );
+      mTable->scrollToItem( item );
+
+
+      item->setText( 0, QString::number( lastDistance() ) );
+      mTotal += lastDistance();
+      editTotal->setText( QString::number( mTotal ) );
+    }
+  }
+  else
+  {
+    editTotal->setText( QString::number( mTotal ) );
+  }
+}
+
+double Qgs3DMeasureDialog::lastDistance()
+{
+  QgsPoint lastPoint = mTool->points().rbegin()[0];
+  QgsPoint secondLastPoint = mTool->points().rbegin()[1];
+  // Euclidean distance
+  qInfo() << "Last line " << lastPoint.x() << lastPoint.y() << lastPoint.z();
+  qInfo() << "Last line " << secondLastPoint.x() << secondLastPoint.y() << secondLastPoint.z();
+  return qSqrt(
+           ( lastPoint.x() - secondLastPoint.x() ) * ( lastPoint.x() - secondLastPoint.x() ) +
+           ( lastPoint.y() - secondLastPoint.y() ) * ( lastPoint.y() - secondLastPoint.y() ) +
+           ( lastPoint.z() - secondLastPoint.z() ) * ( lastPoint.z() - secondLastPoint.z() )
+         );
+}
+
+void Qgs3DMeasureDialog::updateUi()
```

Comment:
This is a LOT of duplicated logic from the 2d case -- can we refactor this to avoid the added tech debt?